### DISCUSSION TITLE
Upgrading to pydantic v2 changes OpenAPI from using `required` to using anyOf with `null`

### DISCUSSION THREAD

**Discussion by rossmacarthur** [fastapi/fastapi#9900] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional
from fastapi import FastAPI
from pydantic import BaseModel

app = FastAPI()


class RequestModel(BaseModel):
    name: str
    age: Optional[int]


@app.get("/")
def index(req: RequestModel):
    return {"message": "Hello World"}
```

### Description

If this is a pydantic v2 issue then I am happy to open an issue there, but I wasn't sure.

When using **pydantic v1** it returns this OpenAPI schema for `RequestModel`

```json
"RequestModel": {
    "properties": {
        "name": {
            "type": "string",
            "title": "Name"
        },
        "age": {
            "type": "integer",
            "title": "Age"
        }
    },
    "type": "object",
    "required": [
        "name"
    ],
    "title": "RequestModel"
},
```

When using **pydantic v2** it returns this OpenAPI schema for `RequestModel`

```json
"RequestModel": {
    "properties": {
        "name": {
            "type": "string",
            "title": "Name"
        },
        "age": {
            "anyOf": [
                {
                    "type": "integer"
                },
                {
                    "type": "null"
                }
            ],
            "title": "Age"
        }
    },
    "type": "object",
    "required": [
        "name",
        "age"
    ],
    "title": "RequestModel"
},
```

I understand that both are correct, however I have code generation that runs off the OpenAPI schema and the former generates better TypeScript code. Is there a way to generate the former?

### Operating System

macOS

### FastAPI Version

0.100.0

### Python Version

3.11.2

**Comment by Kludex** [comment, marked as answer]
https://github.com/pydantic/pydantic/issues/6647
